File: sdk/ruby/smartfield.rb
# ============================================================================
# SmartField Server SDK for Ruby — v2.7.0
# ============================================================================
#
# Copyright (c) 2026 SmartField — MIT License
# Website:  https://smartfield.dev
# Docs:     https://smartfield.dev/docs
# Support:  support@smartfield.dev
#
# Decrypts data encrypted by the <smart-field> browser component.
# Uses Ruby OpenSSL library (RSA-OAEP + AES-256-GCM). Zero external dependencies.
# Keys are generated and stored locally. SmartField never sees your data.
#
# API:
#   sf = SmartField.new
#   sf.init                      — Generate or load RSA-2048 keys
#   sf.get_public_key            — Return public key as JWK hash
#   sf.decrypt(payload)          — Decrypt a single encrypted value
#   sf.decrypt_fields(hash)      — Decrypt all encrypted fields
#
# Encryption: AES-256-GCM (NIST SP 800-38D) + RSA-OAEP-2048 (NIST SP 800-56B)
# ============================================================================
require 'openssl'
require 'json'
require 'base64'
require 'fileutils'

class SmartField
  def initialize
    @private_key = nil
    @public_key = nil
    @public_key_jwk = nil
    @keys_dir = nil
  end

  def init(keys_dir = '.smartfield')
    @keys_dir = keys_dir
    priv_path = File.join(keys_dir, 'private.pem')
    pub_path = File.join(keys_dir, 'public.json')

    if File.exist?(priv_path) && File.exist?(pub_path)
      @private_key = OpenSSL::PKey::RSA.new(File.read(priv_path))
      @public_key = @private_key.public_key
      @public_key_jwk = JSON.parse(File.read(pub_path))
      puts "[SmartField] Keys loaded from #{keys_dir}"
      return
    end

    # Generate new keys
    @private_key = OpenSSL::PKey::RSA.new(2048)
    @public_key = @private_key.public_key

    FileUtils.mkdir_p(keys_dir)

    # Save private key
    File.write(priv_path, @private_key.to_pem)
    File.chmod(0600, priv_path)

    # Save public key as JWK
    @public_key_jwk = key_to_jwk(@public_key)
    File.write(pub_path, JSON.pretty_generate(@public_key_jwk))

    # Add to .gitignore
    gitignore = File.join(Dir.pwd, '.gitignore')
    if File.exist?(gitignore) && !File.read(gitignore).include?('.smartfield')
      File.open(gitignore, 'a') { |f| f.puts "\n# SmartField keys\n.smartfield/" }
    end

    puts "[SmartField] New keys generated in #{keys_dir}"
  end

  def get_public_key
    raise '[SmartField] Not initialized. Call init first.' unless @public_key_jwk
    @public_key_jwk
  end

  def decrypt(encrypted_payload)
    raise '[SmartField] Not initialized. Call init first.' unless @private_key
    return '' if encrypted_payload.nil? || encrypted_payload.empty?

    payload = JSON.parse(Base64.decode64(encrypted_payload))

    iv = Base64.decode64(payload['iv'])
    encrypted_key = Base64.decode64(payload['key'])
    encrypted_data = Base64.decode64(payload['data'])

    # Decrypt AES key with RSA-OAEP
    aes_key_raw = @private_key.private_decrypt(
      encrypted_key,
      OpenSSL::PKey::RSA::PKCS1_OAEP_PADDING
    )

    # Decrypt data with AES-256-GCM
    decipher = OpenSSL::Cipher::AES.new(256, :GCM)
    decipher.decrypt
    decipher.key = aes_key_raw
    decipher.iv = iv

    # GCM tag is last 16 bytes
    tag_length = 16
    decipher.auth_tag = encrypted_data[-tag_length..]
    ciphertext = encrypted_data[0...-tag_length]

    decrypted = decipher.update(ciphertext) + decipher.final
    decrypted
  rescue => e
    raise "[SmartField] Decryption failed: #{e.message}"
  end

  def decrypt_fields(data)
    raise '[SmartField] Not initialized. Call init first.' unless @private_key

    result = {}
    data.each do |key, value|
      if value.is_a?(String) && value.length > 50
        begin
          result[key] = decrypt(value)
        rescue
          result[key] = value
        end
      else
        result[key] = value
      end
    end
    result
  end

  private

  def key_to_jwk(public_key)
    {
      'kty' => 'RSA',
      'alg' => 'RSA-OAEP-256',
      'ext' => true,
      'key_ops' => ['encrypt'],
      'n' => base64_url_encode(public_key.n.to_s(2)),
      'e' => base64_url_encode(public_key.e.to_s(2))
    }
  end

  def base64_url_encode(data)
    Base64.urlsafe_encode64(data).gsub('=', '')
  end
end

File: sdk/python/smartfield.py
"""
============================================================================
SmartField Server SDK for Python — v2.7.0
============================================================================

Copyright (c) 2026 SmartField — MIT License
Website:  https://smartfield.dev
Docs:     https://smartfield.dev/docs
Support:  support@smartfield.dev

Decrypts data encrypted by the <smart-field> browser component.
Keys are generated and stored locally. SmartField never sees your data.

Quick start:
    pip install cryptography

    from smartfield import SmartField
    sf = SmartField()
    sf.init()

    # Flask
    @app.route('/sf-key')
    def key(): return jsonify(sf.get_public_key())

    @app.route('/login', methods=['POST'])
    def login():
        email = sf.decrypt(request.json['email'])
        password = sf.decrypt(request.json['password'])

API:
    sf.init(keys_dir)        — Generate or load RSA-2048 keys
    sf.get_public_key()      — Return public key as JWK dict
    sf.decrypt(payload)      — Decrypt a single encrypted value
    sf.decrypt_fields(dict)  — Decrypt all encrypted fields

Encryption: AES-256-GCM (NIST SP 800-38D) + RSA-OAEP-2048 (NIST SP 800-56B)
============================================================================
"""

import json
import os
import base64
import stat
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.backends import default_backend


class SmartField:
    def __init__(self):
        self._private_key = None
        self._public_key = None
        self._public_key_jwk = None
        self._keys_dir = None

    def init(self, keys_dir='.smartfield'):
        """Generate or load RSA-2048 keys. Keys stored locally, never sent anywhere."""
        self._keys_dir = keys_dir
        priv_path = os.path.join(keys_dir, 'private.pem')
        pub_path = os.path.join(keys_dir, 'public.json')

        if os.path.exists(priv_path) and os.path.exists(pub_path):
            # Load existing keys
            with open(priv_path, 'rb') as f:
                self._private_key = serialization.load_pem_private_key(f.read(), password=None)
            self._public_key = self._private_key.public_key()
            with open(pub_path, 'r') as f:
                self._public_key_jwk = json.load(f)
            print(f'[SmartField] Keys loaded from {keys_dir}')
            return

        # Generate new keys
        self._private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=2048,
            backend=default_backend()
        )
        self._public_key = self._private_key.public_key()

        # Save keys
        os.makedirs(keys_dir, exist_ok=True)

        # Private key (PEM)
        pem = self._private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption()
        )
        with open(priv_path, 'wb') as f:
            f.write(pem)
        os.chmod(priv_path, stat.S_IRUSR | stat.S_IWUSR)  # 0600

        # Public key (JWK)
        self._public_key_jwk = self._key_to_jwk(self._public_key)
        with open(pub_path, 'w') as f:
            json.dump(self._public_key_jwk, f, indent=2)

        # Add to .gitignore
        gitignore = os.path.join(os.getcwd(), '.gitignore')
        if os.path.exists(gitignore):
            with open(gitignore, 'r') as f:
                content = f.read()
            if '.smartfield' not in content:
                with open(gitignore, 'a') as f:
                    f.write('\n# SmartField keys - NEVER commit\n.smartfield/\n')

        print(f'[SmartField] New keys generated in {keys_dir}')

    def get_public_key(self):
        """Return public key as JWK dict. Serve this to your frontend."""
        if not self._public_key_jwk:
            raise Exception('[SmartField] Not initialized. Call init() first.')
        return self._public_key_jwk

    def decrypt(self, encrypted_payload):
        """Decrypt a single SmartField encrypted value."""
        if not self._private_key:
            raise Exception('[SmartField] Not initialized. Call init() first.')
        if not encrypted_payload or not isinstance(encrypted_payload, str):
            return ''

        try:
            payload = json.loads(base64.b64decode(encrypted_payload))

            iv = base64.b64decode(payload['iv'])
            encrypted_key = base64.b64decode(payload['key'])
            encrypted_data = base64.b64decode(payload['data'])

            # Decrypt AES key with RSA
            aes_key_raw = self._private_key.decrypt(
                encrypted_key,
                padding.OAEP(
                    mgf=padding.MGF1(algorithm=hashes.SHA256()),
                    algorithm=hashes.SHA256(),
                    label=None
                )
            )

            # Decrypt data with AES-GCM
            aesgcm = AESGCM(aes_key_raw)
            decrypted = aesgcm.decrypt(iv, encrypted_data, None)

            return decrypted.decode('utf-8')
        except Exception as e:
            raise Exception(f'[SmartField] Decryption failed: {e}')

    def decrypt_fields(self, data):
        """Decrypt all encrypted fields in a dict."""
        if not self._private_key:
            raise Exception('[SmartField] Not initialized. Call init() first.')

        result = {}
        for key, value in data.items():
            if isinstance(value, str) and len(value) > 50:
                try:
                    result[key] = self.decrypt(value)
                except:
                    result[key] = value
            else:
                result[key] = value
        return result

    def _key_to_jwk(self, public_key):
        """Convert RSA public key to JWK format."""
        numbers = public_key.public_numbers()
        e = numbers.e.to_bytes((numbers.e.bit_length() + 7) // 8, 'big')
        n = numbers.n.to_bytes((numbers.n.bit_length() + 7) // 8, 'big')
        return {
            'kty': 'RSA',
            'alg': 'RSA-OAEP-256',
            'ext': True,
            'key_ops': ['encrypt'],
            'e': base64.urlsafe_b64encode(e).rstrip(b'=').decode(),
            'n': base64.urlsafe_b64encode(n).rstrip(b'=').decode()
        }
